### PR TITLE
Fix suspended workers for AsyncLoader

### PR DESCRIPTION
**Pull request by serxa**
This fixes a corner case in deadlock resolution where the job being waited for is prioritized before executing. It led to suspended_workers underflow, which could lead to an unlimited number of workers. Test reproducing the problem is also added.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 7220823689bd49c94eff366c0cc01c5e5a89da73 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58362/7220823689bd49c94eff366c0cc01c5e5a89da73/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58362/7220823689bd49c94eff366c0cc01c5e5a89da73/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Common/AsyncLoader.cpp [hunk 1]
```diff
@@ -691,54 +701,51 @@ void AsyncLoader::wait(std::unique_lock<std::mutex> & job_lock, const LoadJobPtr
         if (worker_pool == job->pool_id)
         {
             job_lock.unlock(); // Avoid reverse locking order
-            workerIsSuspendedByWait(worker_pool, job);
+            std::unique_lock lock{mutex};
             job_lock.lock();
+
+            // Rechecks are required because we have reacquired mutexes
```
Comment (SmitaRKulkarni) [inline src/Common/AsyncLoader.cpp:707]: `std::unique_lock job_lock{job->mutex};` not needed ?

Comment (serxa) [reply src/Common/AsyncLoader.cpp:707]: It also prevents unnecessary unlock/lock that was before the change due to code was inside the function

**Review by SmitaRKulkarni — APPROVED** [review]
LGTM

**Comment by serxa** [comment]
`00191_aggregating_merge_tree_and_final` failure is uinrelated.
Integration tests race seem not related to the changes, but let's run CI once again

**Pull request merged**